stock_picking_on_hold/models/stock.py

```diff
@@ -0,0 +1,212 @@
+        """
+        # self.ensure_one()
+        picking_moves_state_map = defaultdict(dict)
+        picking_move_lines = defaultdict(set)
+        for move in self.env["stock.move"].search([("picking_id", "in", self.ids)]):
+            picking_id = move.picking_id
+            move_state = move.state
+            picking_moves_state_map[picking_id.id].update(
+                {
+                    "any_draft": picking_moves_state_map[picking_id.id].get(
+                        "any_draft", False
+                    )
+                    or move_state == "draft",
+                    "all_cancel": picking_moves_state_map[picking_id.id].get(
+                        "all_cancel", True
+                    )
+                    and move_state == "cancel",
+                    "all_cancel_done": picking_moves_state_map[picking_id.id].get(
+                        "all_cancel_done", True
+                    )
+                    and move_state in ("cancel", "done"),
+                }
+            )
+            picking_move_lines[picking_id.id].add(move.id)
+        for picking in self:
+            picking_id = (picking.ids and picking.ids[0]) or picking.id
+            if not picking_moves_state_map[picking_id]:
+                picking.state = "draft"
+            elif picking_moves_state_map[picking_id]["any_draft"]:
+                picking.state = "draft"
+            elif picking_moves_state_map[picking_id]["all_cancel"]:
+                picking.state = "cancel"
+            elif picking_moves_state_map[picking_id]["all_cancel_done"]:
+                picking.state = "done"
+            else:
+                states = ["confirmed", "assigned", "partially_available"]
+                relevant_move_state = (
+                    self.env["stock.move"]
+                    .browse(picking_move_lines[picking_id])
+                    ._get_relevant_state_among_moves()
+                )
+                if relevant_move_state in states and picking.is_on_hold():
+                    picking.state = "hold"
+                elif picking.immediate_transfer and relevant_move_state not in (
+                    "draft",
+                    "cancel",
+                    "done",
+                ):
+                    picking.state = "assigned"
+                elif relevant_move_state == "partially_available":
+                    picking.state = "assigned"
+                else:
+                    picking.state = relevant_move_state
+
+    dummy_int_to_trigger_state_update = fields.Integer(
+        string="Dummy Integer (to trigger update of state field)"
+    )
+
+    state = fields.Selection(
+        [
+            ("draft", "Draft"),
+            ("waiting", "Waiting Another Operation"),
+            ("confirmed", "Waiting"),
+            ("hold", "Waiting For Payment"),
+            ("assigned", "Ready"),
+            ("done", "Done"),
+            ("cancel", "Cancelled"),
+        ],
```

Comment:
```suggestion
        selection_add = [
            ("hold", "Waiting For Payment"), ("assigned",)
        ], ondelete={'hold': 'set confirmed'}),
```

Reply:
Yes if we use "selection_add" then it will add the state "hold(Waiting For Payment)"  at the last so it was redefined the same state intentionally.

Reply:
No, you didn't understand my suggestion. The trick of adding "("assigned",)" helps in adding "hold" before "assigned". Thus your ordering problem is solved.

Reply:
@MiquelRForgeFlow  stock_picking_on_hold module functionality is picking will be the hold until it is invoice payment and there is another module in the PR which is sale_automatic_workflow functionality is automatically sale order confirmed, the invoice generated and picking is done.

So in this PR, both module functionality is the opposite and the reason to failed the unit test so do you have any idea to fix this?

Reply:
Of course. In https://github.com/OCA/sale-workflow/blob/15.0/.github/workflows/test.yml, do something similar as done in https://github.com/OCA/sale-workflow/blob/14.0/.github/workflows/test.yml#L37L63.

Reply:
@MiquelRForgeFlow Will check and thank you for the help.

Reply:
@MiquelRForgeFlow, Your suggestion regarding excluding the module worked for me.
Thank you